run_as Python Problem

I'm presumably doing something stupid, but I can't seem to actually run a module from Python.

I'm trying to call a module of property type `simde::AOEnergy`. That property type takes a `const chemist::orbital_space::AOSpace<chemist::AOBasisSet<double>>&` and a `const chemist::ChemicalSystem&` as inputs and returns a `double`.

My first Python try:
```.py
import simde

aos = get_an_ao_space()
sys = get_a_chemical_sysetm()
my_pt = simde.simde.AOEnergy
mod = mm.at("Method to run")
e = mod.run_as[my_pt](aos, sys)
```

which results in the error:
```.py
TypeError: Could not find "run_as<simde::AOEnergy>" (set cppyy.set_debug() for C++ errors):
  Failed to instantiate "run_as<simde::AOEnergy>(chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> >&,chemist::ChemicalSystem&)"
```

Then I tried:
```.py
import simde

aos = get_an_ao_space()
sys = get_a_chemical_sysetm()
mod = mm.at("Method to run")
e = mod.run_as["simde::AOEnergy"](aos, sys)
```
Same error. I also tried both of the above with the return specified as `[e]`. I then took the error's advice and set `cppyy.set_debug()`. The ouput:

```
In file included from input_line_177:1:
In file included from /home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/pluginplay.hpp:13:
In file included from /home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type.hpp:2:
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/field_tuple.hpp:316:18: error: constexpr if condition is not a constant expression
    if constexpr(rhs.size() == depth) { // End recursion
                 ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/field_tuple.hpp:357:12: note: in instantiation of function template specialization 'pluginplay::union_guts<pluginplay::FieldTuple<pluginplay::ModuleInput, const chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> > &>, pluginplay::FieldTuple<pluginplay::ModuleInput, const chemist::ChemicalSystem &>, 0>' requested here
    return union_guts(FieldTuple<LHSType, LHSTypes...>(lhs), rhs);
           ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type_impl.hpp:14:33: note: in instantiation of function template specialization 'pluginplay::operator+<pluginplay::ModuleInput, const chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> > &, pluginplay::ModuleInput, const chemist::ChemicalSystem &>' requested here
        return parent.inputs_() + BaseType::inputs();
                                ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type_impl.hpp:31:39: note: in instantiation of member function 'pluginplay::PropertyType<simde::AOEnergy, simde::Energy>::inputs' requested here
    return wrap_(std::forward<T>(rv), inputs(), std::forward<Args>(args)...);
                                      ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/module.hpp:724:32: note: in instantiation of function template specialization 'pluginplay::PropertyType<simde::AOEnergy, simde::Energy>::wrap_inputs<std::map<std::basic_string<char>, pluginplay::ModuleInput, utilities::detail_::CaseInsensitiveLess_, std::allocator<std::pair<const std::basic_string<char>, pluginplay::ModuleInput> > > &, chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> > &, chemist::ChemicalSystem &>' requested here
    temp      = property_type::wrap_inputs(temp, std::forward<Args>(args)...);
                               ^
note: in instantiation of function template specialization 'pluginplay::Module::run_as<simde::AOEnergy, chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> > &, chemist::ChemicalSystem &>' requested here
In file included from input_line_177:1:
In file included from /home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/pluginplay.hpp:13:
In file included from /home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type.hpp:2:
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/field_tuple.hpp:316:18: error: constexpr if condition is not a constant expression
    if constexpr(rhs.size() == depth) { // End recursion
                 ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/field_tuple.hpp:357:12: note: in instantiation of function template specialization 'pluginplay::union_guts<pluginplay::FieldTuple<pluginplay::ModuleResult>, pluginplay::FieldTuple<pluginplay::ModuleResult, double>, 0>' requested here
    return union_guts(FieldTuple<LHSType, LHSTypes...>(lhs), rhs);
           ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type_impl.hpp:24:34: note: in instantiation of function template specialization 'pluginplay::operator+<pluginplay::ModuleResult, pluginplay::ModuleResult, double>' requested here
        return parent.results_() + BaseType::results();
                                 ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type_impl.hpp:59:20: note: in instantiation of member function 'pluginplay::PropertyType<simde::AOEnergy, simde::Energy>::results' requested here
    return unwrap_(results(), std::forward<T>(rv));
                   ^
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/module.hpp:725:27: note: in instantiation of function template specialization 'pluginplay::PropertyType<simde::AOEnergy, simde::Energy>::unwrap_results<std::map<std::basic_string<char>, pluginplay::ModuleResult, utilities::detail_::CaseInsensitiveLess_, std::allocator<std::pair<const std::basic_string<char>, pluginplay::ModuleResult> > > >' requested here
    return property_type::unwrap_results(run(temp));
                          ^
note: in instantiation of function template specialization 'pluginplay::Module::run_as<simde::AOEnergy, chemist::orbital_space::AOSpace<chemist::AOBasisSet<double> > &, chemist::ChemicalSystem &>' requested here
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/module.hpp:725:27: error: no matching member function for call to 'unwrap_results'
    return property_type::unwrap_results(run(temp));
           ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
note: in instantiation of function template specialization 'pluginplay::Module::run_as<simde::AOEnergy>' requested here
/home/ryan/workspaces/NWX/ghostfragment/build/_deps/pluginplay-src/include/pluginplay/property_type/property_type.hpp:140:17: note: candidate template ignored: substitution failure [with T = std::map<std::basic_string<char>, pluginplay::ModuleResult, utilities::detail_::CaseInsensitiveLess_, std::allocator<std::pair<const std::basic_string<char>, pluginplay::ModuleResult> > >]
    static auto unwrap_results(T&& rv);
                ^
```

`FieldTuple` is defined [here](https://github.com/NWChemEx-Project/PluginPlay/blob/master/include/pluginplay/property_type/field_tuple.hpp). The `depth` parameter is a template non-type parameter and `FieldTuple::size()` is constexpr (see [here](https://github.com/NWChemEx-Project/PluginPlay/blob/master/include/pluginplay/property_type/field_tuple.hpp#L141)). 

I'm guessing the `unwrap_results` error is caused by the previous error, but just in case it's not the definition of `unwrap_results` is [here](https://github.com/NWChemEx-Project/PluginPlay/blob/master/include/pluginplay/property_type/property_type_impl.hpp#L42) and I don't see any particular reasonly why substitution would fail unless there was something wrong with the property type.

FTR this is with GCC 11 and the corresponding C++:
```.cpp
simde::type::chemical_system sys(mol, n_electrons);
auto aos = chemist::orbital_space::AOSpace(basis_set);
auto [e] = mod.run_as<simde::AOEnergy>(aos, sys);
```
works fine.

Additional Notes
================

- The C++ definition of the `AOEnergy` class is [here](https://github.com/NWChemEx-Project/SimDE/blob/master/include/simde/energy/ao_energy.hpp).
- The C++ definition of `run_as` is [here](https://github.com/NWChemEx-Project/PluginPlay/blob/master/include/pluginplay/module.hpp#L577).
- A Python example of using `run_as` is [here](https://github.com/NWChemEx-Project/NWChemEx/blob/master/tests/test_python.py#L25)

--- a/include/pluginplay/property_type/field_tuple.hpp
+++ b/include/pluginplay/property_type/field_tuple.hpp
@@ -313,7 +313,7 @@ bool operator!=(const FieldTuple<LHSType, LHSTypes...>& lhs,
  */
 template<typename LHSType, typename RHSType, std::size_t depth = 0>
 auto union_guts(LHSType&& lhs, const RHSType& rhs) {
-    if constexpr(rhs.size() == depth) { // End recursion
+    if constexpr(std::decay_t<RHSType>::size() == depth) { // End recursion
         return lhs;
     } else { // Non-empty RHS
         auto itr = rhs.begin();

